panel/widgets/chat.py

```diff
@@ -0,0 +1,1175 @@
+
+        if len(avatar) == 1:
+            # single character
+            avatar_pane = HTML(avatar)
+        else:
+            try:
+                avatar_pane = Image(avatar, width=35, height=35)
+            except ValueError:
+                # likely an emoji
+                avatar_pane = HTML(avatar)
+        avatar_pane.css_classes = ["avatar"]
+        avatar_pane.visible = self.show_avatar
+        return avatar_pane
+
+    @param.depends("user", "show_user")
+    def _render_user(self) -> None:
+        """
+        Render the user pane as some HTML text or Image pane.
+        """
+        return HTML(self.user, css_classes=["name"], visible=self.show_user)
+
+    @param.depends("value")
+    def _render_value(self):
+        """
+        Renders value as a panel object.
+        """
+        value = self.value
+        if isinstance(value, str):
+            value_panel = Markdown(value)
+        elif isinstance(value, Viewable):
+            value_panel = value
+        else:
+            value_panel = self._create_panel(value)
+        value_panel.css_classes = [*value_panel.css_classes, "message"]
+
+        # used in ChatFeed to extract its contents
+        self._value_panel = value_panel
+        return value_panel
+
+    @param.depends("timestamp", "timestamp_format", "show_timestamp")
+    def _render_timestamp(self) -> None:
+        """
+        Formats the timestamp and renders it as HTML pane.
+        """
+        return HTML(
+            self.timestamp.strftime(self.timestamp_format),
+            css_classes=["timestamp"],
+            visible=self.show_timestamp,
+        )
+
+    def _cleanup(self, root=None) -> None:
+        """
+        Cleanup the exit stack.
+        """
+        if self._exit_stack is not None:
+            self._exit_stack.close()
+            self._exit_stack = None
+        super()._cleanup()
+
+
+class ChatFeed(CompositeWidget):
+    """
+    A widget to display a chat feed with a header and a chat log.
+
+    Parameters
+    ----------
+    value : List[ChatEntry]
+        The entries to add to the chat feed.
+    header : str | Widget
+        The header of the chat feed. Can be a string, pane, or widget.
+    callback : Callable
+        Callback to execute when a user sends a message or
+        when `respond` is called. The signature must include
+        the previous message value `contents`, the previous `user` name,
+        and the component `instance`.
+    callback_user : str
+        The default user name to use for the entry provided by the callback.
+    callback_avatar : str | BinaryIO
+        The default avatar to use for the entry provided by the callback.
+    placeholder : Any
+        Placeholder to display while the callback is running.
+        If not set, defaults to a LoadingSpinner.
+    placeholder_text : str
+        If placeholder is the default LoadingSpinner,
+        the text to display next to it.
+    placeholder_threshold : int
```

Comment:
I would recommend changing the `placeholder_threshold` type to float. Seconds is too limited.

Reply:
Oh I believe it's float. I think I just incorrectly documented it